src/main/java/org/hydev/mcpm/client/database/DatabaseInteractor.java

```diff
@@ -88,6 +93,69 @@ public ListPackagesResult list(ListPackagesInput input) {
         );
     }
 
+    /**
+     * Searches for plugins based on the provided name, keyword, or command.
+     * The input contains the type of search.
+     *
+     * @param input Record of inputs as provided in SearchPackagesInput. See it for more info.
+     * @return Packages result. See the SearchPackagesResult record for more info.
+     */
+    public SearchPackagesResult search(SearchPackagesInput input) {
+        var database = fetcher.fetchDatabase(!input.noCache(), listener);
+
+        if (database == null) {
+            return SearchPackagesResult.by(SearchPackagesResult.State.FAILED_TO_FETCH_DATABASE);
+        }
+
+        var searchStr = input.searchStr();
+        if (searchStr.isEmpty())
+            return SearchPackagesResult.by(SearchPackagesResult.State.INVALID_INPUT);
+
+        var plugins = database.plugins();
+        if (input.type() == SearchPackagesInput.Type.BY_NAME) {
+            return new SearchPackagesResult(
+                    SearchPackagesResult.State.SUCCESS,
+                    searchByName(plugins, searchStr));
+        }
+        else if (input.type() == SearchPackagesInput.Type.BY_COMMAND) {
+            return new SearchPackagesResult(
+                    SearchPackagesResult.State.SUCCESS,
+                    searchByCommand(plugins, searchStr));
+        }
+        else if (input.type() == SearchPackagesInput.Type.BY_KEYWORD) {
+            return new SearchPackagesResult(
+                    SearchPackagesResult.State.SUCCESS,
+                    searchByKeyword(plugins, searchStr));
+        } else {
+            return SearchPackagesResult.by(SearchPackagesResult.State.INVALID_SEARCH_TYPE);
```

Comment:
Nice! If you put this in a valued switch case you don't need to worry about having an invalid switch, there will be a compile error if your switch isn't exhaustive (I think):

```java
switch (input.type()) {
  case SearchPackagesInput.Type.BY_NAME -> { ... }
  case SearchPackagesInput.Type.BY_COMMAND -> { ... }
  case SearchPackagesInput.Type.BY_KEYWORD -> { ... }
}
```

Reply:
switch is definitely cleaner. thx. what should it return by default though?